test/object-store/sync/app.cpp

```diff
@@ -1522,6 +1522,97 @@ TEST_CASE("app: mixed lists with object links", "[sync][app]") {
+    local_config.cache = true;
+    local_config.schema_version = 1;
+    local_config.schema = schema;
+    {
+        auto r = Realm::get_shared_realm(local_config);
+        auto origin = r->read_group().get_table("class_origin");
+        auto target = r->read_group().get_table("class_target");
+        auto other_origin = r->read_group().get_table("class_other_origin");
+        auto other_target = r->read_group().get_table("class_other_target");
+
+        r->begin_transaction();
+        auto o = target->create_object_with_primary_key("Foo").set("name", "Egon");
+        origin->create_object_with_primary_key(47).set("link", o.get_key());
+        other_target->create_object_with_primary_key(UUID("3b241101-e2bb-4255-8caf-4136c566a961"));
+        other_origin->create_object_with_primary_key(ObjectId::gen());
+        r->commit_transaction();
+    }
+
+    auto server_app_config = minimal_app_config(base_url, "upgrade_from_local", schema);
+    auto app_session = create_app(server_app_config);
+    auto app_config = get_config(instance_of<SynchronousTestTransport>, app_session);
+    auto partition = random_string(100);
+    TestSyncManager sync_manager(app_config, {});
+    auto app = sync_manager.app();
+
+    create_user_and_log_in(app);
+    auto user1 = app->current_user();
+    SyncTestFile config(user1, partition, schema);
+
+    auto r1 = Realm::get_shared_realm(config);
+
+    auto origin = r1->read_group().get_table("class_origin");
+    auto target = r1->read_group().get_table("class_target");
+    auto other_origin = r1->read_group().get_table("class_other_origin");
+    auto other_target = r1->read_group().get_table("class_other_target");
+
+    r1->begin_transaction();
+    auto o = target->create_object_with_primary_key("Baa").set("name", "Børge");
+    origin->create_object_with_primary_key(47).set("link", o.get_key());
+    other_target->create_object_with_primary_key(UUID("01234567-89ab-cdef-edcb-a98765432101"));
+    other_origin->create_object_with_primary_key(ObjectId::gen());
+    r1->commit_transaction();
+    CHECK(!wait_for_upload(*r1));
+
+    create_user_and_log_in(app);
+    auto user2 = app->current_user();
+    REQUIRE(user1 != user2);
+
+    local_config.sync_config = std::make_shared<SyncConfig>(user2, bson::Bson(partition));
+    auto new_location = app->sync_manager()->path_for_realm(*local_config.sync_config);
+    util::File::copy(local_config.path, new_location);
+    local_config.path = new_location;
+
+    auto r2 = Realm::get_shared_realm(local_config);
+
+    CHECK(!wait_for_download(*r2));
+    advance_and_notify(*r2);
+    Group& g = r2->read_group();
+    // g.to_json(std::cout);
+    REQUIRE(g.get_table("class_origin")->size() == 1);
+    REQUIRE(g.get_table("class_target")->size() == 2);
+    REQUIRE(g.get_table("class_other_origin")->size() == 2);
+    REQUIRE(g.get_table("class_other_target")->size() == 2);
+
+    CHECK(!wait_for_download(*r1));
+    advance_and_notify(*r1);
```

Comment:
This is a good test. Were you going to check the state of r1 after this as well?
I'd suggest also making another test which converts a more complex Realm from local to sync, and have another user download it without making any modifications, then you can `REQUIRE(original_local_group == second_sync_group)` to make sure that the generated history got all of the state correct without having to manually check each property in the test code. (not blocking)